### DISCUSSION TITLE
Detecting blackholing of long-lived HTTP/3 connections

### DISCUSSION THREAD

**Discussion by amirlivneh** [ngtcp2/ngtcp2#1293] [General]
We have a use case that requires maintaining a long-lived HTTP/3 connection between a mobile app and the server. The data transmission over this connection is sparse, with periods of inactivity lasting up to several minutes. To prevent the connection from closing, we need to use a long idle timeout.

However, we have observed an issue where these connections sometimes become blackholed. We believe this may be caused by different reasons, such as NAT state dropping on the network path. This manifests in sent packets never being acknowledged. Due to the long idle timeout, it takes ngtcp2 a significant amount of time to close the unhealthy connection and notify the app, which can then establish a new connection. The RFC doesn’t seem to address this use case.

To address this, we are considering patching our version of ngtcp2 and introducing a new "ack timeout". The connection will close if there are packets awaiting acknowledgment for more than the specified timeout value. The new timeout will be set to a much lower value that the idle timeout.

Is there a better solution for this problem, ideally one that doesn't require a patch?

**Comment by tatsuhiro-t** [comment]
So you need long timeout when there is no in-flight packet, and but needs another shorter timeout when there is an in-flight packet.
I wonder TCP has similar option for this use case.  In general TCP application has no control over this situation.
I guess that RFC does not specify this because QUIC full handshake needs 1 RTT in contrast to 2 RTT for TCP + TLS, and has true 0-RTT data transfer.  Actively closing a connection and establish one on demand would be the idea behind RFC authors.

**Reply by amirlivneh** [reply]
> I wonder TCP has similar option for this use case. 

[`TCP_USER_TIMEOUT`](https://tinyurl.com/5yuehhd5) may be the TCP equivalent.

> Actively closing a connection and establish one on demand would be the idea behind RFC authors.

Establishing a connection on-demand does not support the scenario where the server initiates data transfer to the client.

**Reply by tatsuhiro-t** [reply]
Thank you for a pointer.  Having TCP option suggests that it has some usecase in TCP application.  In TCP, we can catch error on write(2) failure, but in QUIC just packets are consumed in void and it is in more difficult situation than TCP.  This kind of functionality may be useful if it solves your issue.
